http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java

```diff
@@ -18,139 +18,211 @@
+ * Internal implementation of {@link SdkHttpFullRequest}, buildable via {@link SdkHttpFullRequest#builder()}. Provided to HTTP
+ * implementation to execute a request.
  */
 @SdkInternalApi
+@Immutable
 class DefaultSdkHttpFullRequest implements SdkHttpFullRequest {
-
-    private final Map<String, List<String>> headers;
-    private final String resourcePath;
+    private final String protocol;
+    private final String host;
+    private final Integer port;
+    private final String path;
     private final Map<String, List<String>> queryParameters;
-    private final URI endpoint;
     private final SdkHttpMethod httpMethod;
+    private final Map<String, List<String>> headers;
     private final InputStream content;
 
     private DefaultSdkHttpFullRequest(Builder builder) {
-        this.headers = deepUnmodifiableMap(builder.headers, () -> new TreeMap<>(String.CASE_INSENSITIVE_ORDER));
+        this.protocol = standardizeProtocol(builder.protocol);
+        this.host = Validate.paramNotNull(builder.host, "host");
+        this.port = standardizePort(builder.port);
+        this.path = standardizePath(builder.path);
         this.queryParameters = deepUnmodifiableMap(builder.queryParameters, () -> new LinkedHashMap<>());
-        this.resourcePath = builder.resourcePath;
-        this.endpoint = builder.endpoint;
-        this.httpMethod = builder.httpMethod;
+        this.httpMethod = Validate.paramNotNull(builder.httpMethod, "method");
+        this.headers = deepUnmodifiableMap(builder.headers, () -> new TreeMap<>(String.CASE_INSENSITIVE_ORDER));
         this.content = builder.content;
     }
 
+    private String standardizeProtocol(String protocol) {
+        Validate.paramNotNull(protocol, "protocol");
+
+        String standardizedProtocol = StringUtils.lowerCase(protocol);
+        Validate.isTrue(standardizedProtocol.equals("http") || standardizedProtocol.equals("https"),
+                        "Protocol must be 'http' or 'https', but was %s", protocol);
+
+        return standardizedProtocol;
+    }
+
+    private String standardizePath(String path) {
+        if (StringUtils.isEmpty(path) || path.equals("/")) {
+            return "";
+        }
+
+        StringBuilder standardizedPath = new StringBuilder();
+
+        // Path must always start with '/'
+        if (!path.startsWith("/")) {
+            standardizedPath.append('/');
+        }
+
+        standardizedPath.append(path);
+
+        // Path must never end with '/', unless they're an empty path.
+        if (standardizedPath.length() > 1 && path.endsWith("/")) {
+            standardizedPath.setLength(standardizedPath.length() - 1);
+        }
+
+        return standardizedPath.toString();
+    }
+
+    private Integer standardizePort(Integer port) {
+        Validate.isTrue(port == null || port >= -1,
+                        "Port must be positive (or null/-1 to indicate no port), but was '%s'", port);
+
+        if (port != null && port == -1) {
```

Comment:
Do you need the null check?

Reply:
Yeah, if port is null, the unboxing involved in port == -1 would throw a NullPointerException.